### PR TITLE
Register and Remove Action Manager and Tests

### PR DESCRIPTION
**Pull request by caitlynachen**

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+		intent = [lux.Clause("?",data_type="nominal")]
+		vlist = VisList(intent,ldf)
+		for vis in vlist:
+			vis.score = 10
+		vlist = vlist.topK(15)
+		return {"action":"bars", "description": "Random list of Bar charts", "collection": vlist}
+	lux.register_action("bars", random_categorical)
+	df._repr_html_()
+	assert(len(df.recommendation["bars"]) > 0)
+	assert("bars" in df.recommendation)
```
Comment (dorisjlee) [inline tests/test_config.py:88]: Can we share the code used inside q1, q2, q3, q7, q9 tests? There is a lot of duplicated code across the test cases.

File: tests/test_config.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		vlist = vlist.topK(15)
+		return {"action":"bars", "description": "Random list of Bar charts", "collection": vlist}
+	def contain_horsepower(df):
+		for clause in df.intent:
+			if clause.get_attr() == "Horsepower":
+				return True
+		return False
+	lux.register_action("bars", random_categorical, contain_horsepower)
+	df._repr_html_()
+	assert("bars" not in df.recommendation)
```
Comment (dorisjlee) [inline tests/test_config.py:54]: Add reason for assertion fail, i.e. "The action `bars` should not be rendered since intent does not satisfy `contain_horsepower`."

File: lux/_config/config.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+actions = DictWrapper(_registered_actions)
+
+
+def register_action(
+    key: str = "",
+    function: Optional[Callable[[Any], Any]] = None,
+    validator: Optional[Callable[[Any], Any]] = None,
+    *args,
+) -> None:
+
```
Comment (dorisjlee) [inline lux/_config/config.py:71]: Add docstring

File: lux/_config/config.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	def __len__(self):
+		return len(list(self.d.keys()))
+
+	def __dir__(self) -> Iterable[str]:
+		return list(self.d.keys())
+
+actions = DictWrapper(_registered_actions)
+
+
+def register_action(
```
Comment (dorisjlee) [inline lux/_config/config.py:65]: Can we rename these arguments as something else? 
`key` --> `name`
`function` --> `action`
`validator` --> This is a bit duplicate to the Lux Validator, which is something completely different. We can name this as display condition or something more descriptive, e.g., `display_if`, `display_condition`, `validate_condition`, `applicable_if`, etc.

File: lux/_config/config.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    function: Optional[Callable[[Any], Any]] = None,
+    validator: Optional[Callable[[Any], Any]] = None,
+    *args,
+) -> None:
+
+	key = key.lower()
+	if function:
+		is_callable(function)
+	if not function:
+		raise ValueError("No parameter function found")
```
Comment (dorisjlee) [inline lux/_config/config.py:76]: Make more descriptive.

File: lux/_config/config.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+	def __dir__(self) -> Iterable[str]:
+		return list(self.d.keys())
+
+actions = DictWrapper(_registered_actions)
+
+
+def register_action(
+    key: str = "",
+    function: Optional[Callable[[Any], Any]] = None,
```
Comment (dorisjlee) [inline lux/_config/config.py:67]: Why is `function` an optional argument? It should be required?
Should we enforce the output type of the Callable be a list of VisList and input type as something more specific?

File: lux/_config/config.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+	key = key.lower()
+	if key not in _registered_actions:
+		raise ValueError(f"Option '{key}' has not been registered")
+
+	del _registered_actions[key]
+	update_actions["flag"] = True
+
+def is_callable(obj) -> bool:
+	"""
```
Comment (dorisjlee) [inline lux/_config/config.py:97]: indentation off

File: lux/action/custom.py [hunk 8]
```diff
@@ -42,4 +43,20 @@ def custom(ldf):
...
     vlist.sort(remove_invalid=True)
-    return recommendation
\ No newline at end of file
+    return recommendation
+
+def custom_actions(ldf):
+    if (lux.actions.__len__() > 0):
+        recommendations = []
+        for action_name in lux.actions.__dir__():
+            validator = lux.actions.__getattr__(action_name).validator
```
Comment (dorisjlee) [inline lux/action/custom.py:52]: What is the difference between `_get_action`, ` __getactions__`, and `__getattr__(action_name)`?

File: lux/_config/config.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+'''
+from collections import namedtuple
+from typing import Any, Callable, Dict, Iterable, List, Optional
+
+RegisteredOption = namedtuple("RegisteredOption", "key function validator args")
+
+# holds registered option metadata
+_registered_actions: Dict[str, RegisteredOption] = {}
+update_actions: Dict[str, bool] = {}
+update_actions["flag"] = False
```
Comment (dorisjlee) [inline lux/_config/config.py:13]: Can you add a comment to explain what `update_actions` does?

File: lux/core/frame.py [hunk 10]
```diff
@@ -695,7 +699,7 @@ def rec_to_JSON(recs):
 				# delete DataObjectCollection since not JSON serializable
 				del rec_lst[idx]["collection"]
 		return rec_lst
-
```
Comment (dorisjlee) [inline lux/core/frame.py:698]: Remove unnecessary deletion and addition of whitespace and newlines throughout the PR.

File: tests/test_config.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+		for clause in df.intent:
+			if clause.get_attr() == "Horsepower":
+				return True
+		return False
+	df = pd.read_csv("lux/data/car.csv")
+	lux.register_action("bars", random_categorical, contain_horsepower)
+	df.set_intent(["Acceleration", "Horsepower"])
+	df._repr_html_()
+	assert("bars" in df.recommendation)
+	assert(len(df.recommendation["bars"]) > 0)
```
Comment (dorisjlee) [inline tests/test_config.py:179]: This is a really comprehensive test suite that tests many of the edge cases of how the `register_action` and `remove_action` might be used!

File: tests/test_config.py (hunk 1 above)
Comment (dorisjlee) [reply tests/test_config.py:88]: Can we remove the q* in the test name? The test assertions should be description so that when the test fails, it would be self-explanatory. We can add these text by doing `assert( CONDITION, TEXT_DISPLAYED_IF_FAILED)`.

File: lux/_config/config.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+		try:
+			v = object.__getattribute__(self, "d")[key]
+		except KeyError as err:
+			raise OptionError("No such option") from err
+		if isinstance(v, dict):
+			return DictWrapper(v, prefix)
+		else:
+			return _get_action(prefix)
+
+	def __getactions__(self):
```
Comment (dorisjlee) [inline lux/_config/config.py:50]: Add docstring and return type.

**Review by dorisjlee — CHANGES_REQUESTED** [review]

**Comment by dorisjlee** [comment]
Thanks @caitlynachen. The commit looks great. Can you add docstrings to the newly added functions and config, as well as a description for this PR? I've added some suggested changes and questions in the code review.

File: lux/action/custom.py (hunk 8 above)
Comment (caitlynachen) [reply lux/action/custom.py:52]: __getactions__ returns a list of actions. _get_action is unused so I will probably delete it

File: lux/_config/config.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+class DictWrapper:
+	def __init__(self, d: Dict[str, Any], prefix: str = ""):
+		object.__setattr__(self, "d", d)
+		object.__setattr__(self, "prefix", prefix)
+	def __init__(self, d: Dict[str, RegisteredOption], prefix: str = ""):
+		object.__setattr__(self, "d", d)
+		object.__setattr__(self, "prefix", prefix)
+
+	def __getattr__(self, name: str):
```
Comment (dorisjlee) [inline lux/_config/config.py:26]: Can you fix the indentation issue in the docstring and throughout?

File: lux/_config/config.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    *args,
+) -> None:
+	"""
+    Parameters
+    ----------
+    `name` - the id for the actions
+    `action` - the function to be applied to the dataframe
+    `display_condition` - the function to check whether or not the function should be applied
+    `args` - any additional arguments the function may require
+    Description
```
Comment (dorisjlee) [inline lux/_config/config.py:83]: `Description` doesn't need header and goes before `Parameters.

File: lux/_config/config.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	name: str = "",
+    action: Callable[[Any], Any] = None,
+    display_condition: Optional[Callable[[Any], Any]] = None,
+    *args,
+) -> None:
+	"""
+    Parameters
+    ----------
+    `name` - the id for the actions
+    `action` - the function to be applied to the dataframe
```
Comment (dorisjlee) [inline lux/_config/config.py:80]: Change to `action - the function used to generate the recommendations`

File: lux/_config/config.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+	update_actions["flag"] = True
+
+def remove_action(
+	name: str = "",
+) -> None:
+	"""
+    Parameters
+    ----------
+    `name` - the id for the action to remove
+    Description
```
Comment (dorisjlee) [inline lux/_config/config.py:105]: `Description` doesn't need header and goes before `Parameters`. The docstring in your`__getattr__` is a good example.

File: lux/_config/config.py (hunk 7 above)
Comment (dorisjlee) [reply lux/_config/config.py:117]: indentation still a bit off needs to align with first line of code.

File: lux/_config/config.py (hunk 16 above)
Comment (dorisjlee) [reply lux/_config/config.py:105]: Indentation off .

File: lux/core/frame.py [hunk 17]
```diff
@@ -224,12 +225,10 @@ def set_intent(self, intent:List[Union[str, Clause]]):
 		"""
 		Main function to set the intent of the dataframe.
 		The intent input goes through the parser, so that the string inputs are parsed into a lux.Clause object.
-
```
Comment (dorisjlee) [inline lux/core/frame.py:227]: Add back these spacing.

File: lux/core/frame.py [hunk 18]
```diff
@@ -468,7 +473,6 @@ def widget(self):
 	def exported(self) -> Union[Dict[str,VisList], VisList]:
 		"""
 		Get selected visualizations as exported Vis List
-
```
Comment (dorisjlee) [inline lux/core/frame.py:471]: Add back in the newline.

File: lux/_config/config.py [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+	def __init__(self, d: Dict[str, RegisteredOption], prefix: str = ""):
+		object.__setattr__(self, "d", d)
+		object.__setattr__(self, "prefix", prefix)
+
+	def __getattr__(self, name: str):
+		"""
+    	Gets a specific registered action by id
+    	Parameters
+    	----------
+    	`name` - the id for the actions
```
Comment (dorisjlee) [inline lux/_config/config.py:31]: Change to: `name - the name of the action`. Indentation is off from line 28-35.

File: lux/_config/config.py (hunk 9 above)
Comment (dorisjlee) [reply lux/_config/config.py:13]: Still unclear what `update_actions` does, I think it checks if the actions are updated? Adding a line comment here would be useful.

**Review by dorisjlee — CHANGES_REQUESTED** [review]

**Comment by dorisjlee** [comment]
Added some minor comments on the indentation, it looks good overall and should be ready to be merged after this round of fixes. Thanks!

File: lux/_config/config.py (hunk 9 above)
Comment (caitlynachen) [reply lux/_config/config.py:13]: my bad I had to revert some comments during editing to test ill add it back

**Review by dorisjlee — APPROVED** [review]

**Comment by dorisjlee** [comment]
This looks great, thanks @caitlynachen !
